### PR TITLE
Isin extension

### PR DESCRIPTION
**Pull request by pillarxyz**
#5716 

added most backend implementations there is only problem with tensorflow I'm still trying to solve since it doesnt have the function isin, once I'm able to do that I will add tests

### CODE REVIEW

**Comment by pillarxyz** [comment]
I'm really sorry for this mess of a pull request, I'm still not that good with working with git

just to summerize what I did here

- Added isin for all backends
- Created the implementation for ivy and added wrapping for the array class
- Added the tests but I think I didnt do it well since I'm not familiar with how tests working for the project

**Comment by abdrahmandiab** [comment]
Hi @pillarxyz ,
As Saeed is taking a hiatus from working with our team, I will be taking over the reviewing process for this PR!
Regarding the messy commits, that's definitely not a problem! We will be squashing this all into one commit while merging so no worries there! 

I have self-requested a review for now as per your last comment which indicates the PR is ready for review, but in the future please request this review yourself once the code is ready. 😄 

Thanks and please be patient with the review as I have quite a few to go through now! 😅

**Comment by pillarxyz** [comment]
@abdrahmandiab all good, thanks alot

**Review by abdrahmandiab — CHANGES_REQUESTED** [review]
Hi @pillarxyz , 
It seems that the tests for the function `isin` you implemented are currently failing because there are two parameters missing `with_out` and `container_flags` from the `helpers.test_function` method used for testing. 
For more information you can check the CI test logs [here](https://github.com/unifyai/ivy/actions/runs/3274977034/jobs/5389231390#step:6:517). 

For future reference, you can access the ci test logs by clicking on the `Details` button next to the failing check in our CI tests.
image

After this you can check for the specific error by going to the `Run Extensions Test` section and searching the logs using the top right search bar for `isin`, example shown below... 😅
image

Could you please work on fixing these errors? 😄
Thanks so much for your work so far, and please don't hesitate to ask for help or re-request a review! 🚀

**Comment by pillarxyz** [comment]
I will fix it right away, I have a question tho since I noticed that we now have an extensions folder with seperate files where do you see this function fitting ? my first guess was it would be in array/extensions/general.py or array/extensions/utility.py

**Comment by abdrahmandiab** [comment]
> I will fix it right away, I have a question tho since I noticed that we now have an extensions folder with seperate files where do you see this function fitting ? my first guess was it would be in array/extensions/general.py or array/extensions/utility.py

Hey @pillarxyz ,
sorry for the late reply 😅

Yes you can go ahead and implement this in the `extensions/general.py` file for now!
Thanks!

**Comment by pillarxyz** [comment]
> > I will fix it right away, I have a question tho since I noticed that we now have an extensions folder with seperate files where do you see this function fitting ? my first guess was it would be in array/extensions/general.py or array/extensions/utility.py
> 
> Hey @pillarxyz , sorry for the late reply 😅
> 
> Yes you can go ahead and implement this in the `extensions/general.py` file for now! Thanks!

Hey @abdrahmandiab I already done that, can you check now if the PR is ready to be merged

**Comment by pillarxyz** [comment]
Hello @abdrahmandiab 
Mind checking the PR, if there is any modification to be done or is it ready for merge ?

**Comment by abdrahmandiab** [comment]
> Hello @abdrahmandiab Mind checking the PR, if there is any modification to be done or is it ready for merge ?

Hi @pillarxyz , super sorry for the late review, been quite swamped lately! I will get to this within today or tomorrow hopefully so be on the lookout for that!

Thanks! and apologies for the inconvenience! 😅

File: ivy_tests/test_ivy/test_functional/test_extensions/test_core/test_general.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        native_array_flags=native_array,
+        instance_method=instance_method,
+        container_flags=container,
+        fw=fw,
+        fn_name="isin",
+        ground_truth_backend='numpy',
+        elements=elements,
+        test_elements=test_elements,
+        invert=invert,
+        assume_unique=assume_unique,
```
Comment (abdrahmandiab) [inline ivy_tests/test_ivy/test_functional/test_extensions/test_core/test_general.py:47]: The `assume_unique` parameter seems to produce unpredictable/incorrect behavior when set to `True` when `fw` is `'tensorflow'`. 😅

**Review by abdrahmandiab — CHANGES_REQUESTED** [review]
Hi @pillarxyz ,
Great work so far, and apologies once again for the late review 😅 
I have ran the test that you have implemented and found that it still throws some errors. I've already fixed most of the trivial ones which were causing some of the backends to fail, and you can see the changes I've in the commits.

Unfortunately however there still seems to be a problem with the behavior of the parameter `assume_unique` when it comes to the `tensorflow` backend implementation. Could you please try fixing that? and if you face any trouble you can ask me and I will hopefully reply faster this time for sure! 😅

File: ivy_tests/test_ivy/test_functional/test_extensions/test_core/test_general.py (hunk 1 above)
Comment (abdrahmandiab) [reply ivy_tests/test_ivy/test_functional/test_extensions/test_core/test_general.py:47]: > Hi @pillarxyz , Great work so far, and apologies once again for the late review 😅 I have ran the test that you have implemented and found that it still throws some errors. I've already fixed most of the trivial ones which were causing some of the backends to fail, and you can see the changes I've made in the commits.
> 
> Unfortunately however there still seems to be a problem with the behavior of the parameter `assume_unique` when it comes to the `tensorflow` backend implementation. Could you please try fixing that? and if you face any trouble you can ask me and I will hopefully reply faster this time for sure! 😅
> 
> Thanks!

One more note, as this bug is kind of flaky and not very easy to reproduce, if you are unable to reproduce it we can just merge the PR and handle it later if it starts failing :)

**Comment by abdrahmandiab** [comment]
One more note, as this bug is kind of flaky and not very easy to reproduce, if you are unable to reproduce it we can just merge the PR and handle it later from our side if it starts failing :)

**Comment by pillarxyz** [comment]
> One more note, as this bug is kind of flaky and not very easy to reproduce, if you are unable to reproduce it we can just merge the PR and handle it later from our side if it starts failing :)

Yeah I've been having a hard time with the tensorflow implementation. I don't know if I'll be able to work on it the next few days since I'm getting my laptop fixed (I dropped coffee on it 😅)

**Comment by pillarxyz** [comment]
> > One more note, as this bug is kind of flaky and not very easy to reproduce, if you are unable to reproduce it we can just merge the PR and handle it later from our side if it starts failing :)
> 
> Yeah I've been having a hard time with the tensorflow implementation. I don't know if I'll be able to work on it the next few days since I'm getting my laptop fixed (I dropped coffee on it 😅) 

@abdrahmandiab Also thanks a lot for the help in this PR, its been great.

**Comment by abdrahmandiab** [comment]
> > > 
> @abdrahmandiab Also thanks a lot for the help in this PR, its been great.

it's my pleasure! truly, and sorry once again for the late reviews haha!
Also, shame about the laptop 😅 Happens to the best of us! Hope it will be fixed soon!

**Comment by pillarxyz** [comment]
I dont how to deal with these .idea folders getting added to the commit after I solve merge conflicts😅

File: ivy/container/experimental/image.py [hunk 2]
```diff
@@ -1,5 +1,5 @@
 from ivy.container.base import ContainerBase
 
 
-class ContainerWithImageExperimental(ContainerBase):
+class ContainerWithGeneralExtensions(ContainerBase):
```
Comment (abdrahmandiab) [inline ivy/container/experimental/image.py:4]: why was this change made? 😅

**Comment by abdrahmandiab** [comment]
> I dont how to deal with these .idea folders getting added to the commit after I solve merge conflicts😅

Hi @pillarxyz, could you tell me what method you use for committing things? is it the terminal or through your IDE `source control` window? also what IDE do you use? 😅

**Comment by pillarxyz** [comment]
> Hi @pillarxyz, could you tell me what method you use for committing things? is it the terminal or through your IDE source control window? also what IDE do you use? 😅

I'm using pycharm with ivy conda environment, and using git in the terminal

File: ivy/container/experimental/image.py (hunk 2 above)
Comment (pillarxyz) [reply ivy/container/experimental/image.py]: but I fixed that right now

**Comment by pillarxyz** [comment]
@abdrahmandiab I think that the tensorflow implementation works fine now, although a particular example keeps failing for some reason even though when I try the function in the python console with the inputs in the examples it delivers the expected results.

**Comment by pillarxyz** [comment]
Hello again

I saw these in both the numpy and jax implementations of isin

image

image

which suggests that this argument is already problematic in both of those frameworks.

**Comment by abdrahmandiab** [comment]
hi @pillarxyz,
as a last effort, I think the `assume_unique` output might be producing incorrect results because the generated arrays in these specific test cases themselves are not unique.

So I think before we give up and merge it, if you could edit the data generation to ensure that the generated arrays contain unique elements, I think this might stop the failing examples. My initial idea to do this is to use something like...
```
dtypes_and_values(...).filter( 
    lambda x: x == np.unique(x) #np.unique returns the unique values in an array
)
```

**Comment by abdrahmandiab** [comment]
> I'm using pycharm with ivy conda environment, and using git in the terminal

Aha, well in this case where you are using it from the terminal, please refrain from using `git add .` and rather please use `git add <specific file name>`  , this might be a bit tedious but it will save you the trouble of committing unwanted changes that are related to the IDE or to the Array API tests. 😅

**Comment by abdrahmandiab** [comment]
Hi @pillarxyz,
after so long working on this PR, we've finally done it! 🚀

Thanks so much for your hard work, and good luck with the rest of your application process! <3

**Comment by pillarxyz** [comment]
> Hi @pillarxyz, after so long working on this PR, we've finally done it! 🚀
> 
> Thanks so much for your hard work, and good luck with the rest of your application process! <3

Thanks alot for the help, I've learnt alot about the project thanks to you <3

**Pull request merged**